File: Assets/Scripts/HookR.cs
using UnityEngine;

public class HookR
{
    
    // parameters
    Transform user;
    Transform userCamera;
    Rigidbody userRigidBody;
    float physicsMultiplier;
    LayerMask ground;
    float hookRange = 125f;
    float hookPower = 50f;

    // state
    LineRenderer lineRenderer1;
    LineRenderer lineRenderer2;
    bool isHooked = false;
    Vector3 hookPos;

    public HookR(GameObject user)
    {
        MainUser script = user.GetComponent<MainUser>();
        this.user = user.transform;
        this.userCamera = script.playerCamera.transform;
        this.userRigidBody = user.GetComponent<Rigidbody>();
        this.physicsMultiplier = script.physicsMultiplier;
        this.ground = script.ground;

        GameObject Hook1 = new GameObject("HookR1");
        GameObject Hook2 = new GameObject("HookR2");
        Hook1.transform.parent = user.transform;
        Hook2.transform.parent = user.transform;
        lineRenderer1 = Hook1.AddComponent<LineRenderer>();
        lineRenderer2 = Hook2.AddComponent<LineRenderer>();

        lineRenderer1.startWidth = 0.1f;
        lineRenderer1.endWidth = 0.1f;
        lineRenderer2.startWidth = 0.1f;
        lineRenderer2.endWidth = 0.1f;
        lineRenderer1.material = new Material(Shader.Find("Sprites/Default"));
        lineRenderer1.startColor = Color.blue;
        lineRenderer1.endColor = Color.blue;
        lineRenderer2.material = new Material(Shader.Find("Sprites/Default"));
        lineRenderer2.startColor = Color.blue;
        lineRenderer2.endColor = Color.blue;
    }

    public void Start()
    {
        RaycastHit hit;
        if(Physics.Raycast(userCamera.position, userCamera.forward, out hit, hookRange, layerMask: ground))
        {
            hookPos = hit.point;
            Debug.Log("Hook Right: from "+userCamera.position+" to "+hookPos+"");
            isHooked = true;
            Vector3 startForce = (hookPos - user.position).normalized * hookPower * physicsMultiplier;
            if (startForce.y < physicsMultiplier * hookPower * 0.5f) startForce += Vector3.up * (physicsMultiplier * hookPower * 0.5f - startForce.y);
            if (hookPos.y >= user.position.y) startForce += Vector3.up * (hookPos.y - user.position.y) * physicsMultiplier;
            userRigidBody.AddForce(startForce, ForceMode.Force);
            lineRenderer1.enabled = true;
            lineRenderer1.SetPosition(0, user.position + user.right * 0.5f + user.up * 0.5f);
            lineRenderer1.SetPosition(1, hookPos);
            lineRenderer2.enabled = true;
            lineRenderer2.SetPosition(0, user.position + user.right * 0.5f - user.up * 0.5f);
            lineRenderer2.SetPosition(1, hookPos);
        }
    }

    public void Frame()
    {
        if (isHooked)
        {
            lineRenderer1.SetPosition(0, user.position + user.right * 0.5f + user.up * 0.5f);
            lineRenderer2.SetPosition(0, user.position + user.right * 0.5f - user.up * 0.5f);
        }
    }

    public void Tick()
    {
        if (isHooked)
        {
            Vector3 direction = (hookPos - user.position).normalized;
            Vector3 force = direction * hookPower * physicsMultiplier;
            userRigidBody.AddForce(force, ForceMode.Force);
            if (Vector3.Angle(userRigidBody.velocity, force) > 90f) userRigidBody.AddForce(-userRigidBody.velocity * physicsMultiplier);
        }
    }

    public void Stop()
    {
        isHooked = false;
        lineRenderer1.enabled = false;
        lineRenderer2.enabled = false;
    }
}

File: Assets/StreamingAssets/Abilities/HookR.lua
---@diagnostic disable: undefined-global

local userObject
local userCamera
local userRigidBody
local physicsMultiplier
local ground
local hookRange = 125
local hookPower = 50

local lineRenderer1
local lineRenderer2
local isHooked = false
local hookPos

local script = user:GetComponent("MainUser")
userObject = user.transform
userCamera = script.playerCamera.transform
userRigidBody = user:GetComponent("Rigidbody")
physicsMultiplier = script.physicsMultiplier
ground = script.ground

local Hook1 = CreateGameObject("Hook1")
local Hook2 = CreateGameObject("Hook2")
Hook1.transform.parent = user.transform
Hook2.transform.parent = user.transform

lineRenderer1 = AddLineRenderer(Hook1)
lineRenderer2 = AddLineRenderer(Hook2)

lineRenderer1.startWidth = 0.1
lineRenderer1.endWidth = 0.1
lineRenderer2.startWidth = 0.1
lineRenderer2.endWidth = 0.1
lineRenderer1.material = CreateMaterial(Shader.Find("Sprites/Default"))
lineRenderer2.material = CreateMaterial(Shader.Find("Sprites/Default"))
lineRenderer1.startColor = Color.blue
lineRenderer1.endColor = Color.blue
lineRenderer2.startColor = Color.blue
lineRenderer2.endColor = Color.blue
lineRenderer1.enabled = false
lineRenderer2.enabled = false

function Start()
    if RaycastCheck(userCamera.position, userCamera.forward, hookRange, ground) then
        local hit = RaycastValue(userCamera.position, userCamera.forward, hookRange, ground)
        hookPos = hit.point
        isHooked = true
        local startForce = Normalize(hookPos - userObject.position) * hookPower * physicsMultiplier
        if startForce.y < physicsMultiplier * hookPower * 0.5 then
            startForce = startForce + Vector3.up * (physicsMultiplier * hookPower * 0.5 - startForce.y)
        end
        if hookPos.y >= userObject.position.y then
            startForce = startForce + Vector3.up * (hookPos.y - userObject.position.y) * physicsMultiplier
        end
        AddForce(userRigidBody, startForce, "Force")
        lineRenderer1.enabled = true
        lineRenderer1:SetPosition(0, userObject.position + userObject.right * 0.5 + userObject.up * 0.5)
        lineRenderer1:SetPosition(1, hookPos)
        lineRenderer2.enabled = true
        lineRenderer2:SetPosition(0, userObject.position + userObject.right * 0.5 - userObject.up * 0.5)
        lineRenderer2:SetPosition(1, hookPos)
    end
end

function Frame()
    if isHooked then
        lineRenderer1:SetPosition(0, userObject.position + userObject.right * 0.5 + userObject.up * 0.5)
        lineRenderer2:SetPosition(0, userObject.position + userObject.right * 0.5 - userObject.up * 0.5)
    end
end

function Tick()
    if isHooked then
        local direction = Normalize(hookPos - userObject.position)
        local force = direction * hookPower * physicsMultiplier
        AddForce(userRigidBody, force, "Force")
        if Vector3.Angle(userRigidBody.velocity, force) > 90 then
            AddForce(userRigidBody, -userRigidBody.velocity * physicsMultiplier, "Force")
        end
    end
end

function Stop()
    isHooked = false
    lineRenderer1.enabled = false
    lineRenderer2.enabled = false
end